### DISCUSSION TITLE
Suricata events coming with delay

### DISCUSSION THREAD

**Discussion by bit1290** [Security-Onion-Solutions/securityonion#6464] [Unsupported Versions]
Hello,

We are forwarding events from Suricata to and external SIEM and this is our logstash template to get those events to be sent:

![image](https://user-images.githubusercontent.com/16745902/144871215-73fb6d7a-3de7-455b-9811-81d0e01a0087.png)

And this is our logstash configuration file:

[papidss003_standalone.txt](https://github.com/Security-Onion-Solutions/securityonion/files/7661675/papidss003_standalone.txt)

The thing is that some of the logs are reaching the SIEM with a delay of 1/2/3 hours as we can see from the image bellow:

![image](https://user-images.githubusercontent.com/16745902/145028887-3ff6949b-4053-4701-ad5e-ae917ca154b6.png)

As we can see the log arrived to the SIEM at Dec 6, 2021, 3:17:43 PM (Start Time) but it is a log from 2021-12-06T12:09:10PM (timestamp in the log).
So logstash is still retrieving events from old log files, I confirmed that. In this case we have the following files:

![image](https://user-images.githubusercontent.com/16745902/144872500-dcecb0d3-48b5-4e20-af4f-c3491a36b562.png)

And logstash is still retrieving logs from  eve-2021-12-06-13:01.json generated 2 hours ago, however at the same time is retrievieng logs updated.

Could you help us figuring out why this issue happens?

Thanks a lot!

**Comment by bit1290** [comment]
Here is my filebeat configuration file:

[filebeat.txt](https://github.com/Security-Onion-Solutions/securityonion/files/7662206/filebeat.txt)

**Comment by dougburks** [comment]
https://docs.securityonion.net/en/2.3/logstash.html#forwarding-events-to-an-external-destination

**Reply by bit1290** [reply]
I've tried to add the following configuration, but nothing is sent through tcpdump:

[root@papidss003 custom]# pwd
/opt/so/saltstack/local/salt/logstash/pipelines/config/custom
[root@papidss003 custom]# cat suricata2.conf
filter {
if [module] =~ "suricata" {
clone {
id => "clone_zeek_dns_events"
clones => ["zeek-dns-clone"]
add_tag => [ "clone" ]
}
}
}
output {
if "clone" in [tags] {
tcp {
id => "cloned_events_out"
host => "10.135.101.131
port => 514
codec => "json_lines"
}
}
}
[root@papidss003 custom]#


Is there something I'm missing here?

**Comment by bit1290** [comment]
Hello,

I did it with the following configuration:

[root@papidss003 custom]# cat suricata2.conf
filter {
  if [module] =~ "suricata" {
    clone {
        id => "clone_suricata_eve_events"
        clones => ["suricata-eve-clone"]
        add_tag => [ "suricataclone" ]
    }
  }
}
output {
  if "suricataclone" in [tags]{
    tcp {
      id => "cloned_events_out_suricata"
      host => "10.135.101.131"
      port => 514
      codec => "json_lines"
    }
  }
}

But the events are still delaying.

**Reply by weslambert** [reply]
To be clear, are the logs...

- arriving in Security Onion AND the destination SIEM late
or
- arriving in only the destination SIEM late
?

**Reply by bit1290** [reply]
If I do a tcpdump on Security Onion I can see that the events are being sent late to the SIEM:

![image](https://user-images.githubusercontent.com/16745902/146184909-8174c07c-70d0-491c-bde3-8d7a3abf1714.png)

As you can see it is sending events from day 14th and today is 15th and from a log file that should have already been read.

Is there something I'm doing wrong?

Thanks for you reply.

**Reply by weslambert** [reply]
Thanks for the additional information, however, you didn't answer the original question:

To be clear, are the logs...

arriving in Security Onion AND the destination SIEM late
or
arriving in only the destination SIEM late
?

**Reply by bit1290** [reply]
Arriving in only the destination SIEM late.

**Reply by weslambert** [reply]
If events are being ingested into Security Onion in a timely fashion, but not in the destination SIEM, it might suggest that either the pipeline config handling the cloning and output to the SIEM is not cloning/sending fast enough, or the downstream node is not taking in the data fast enough.

Have you performed any investigation on the SIEM side to ensure the events are not being queued/stalled before being indexed, and that Logstash can send at the rate it needs to send? You may also look in the Logstash log in `/opt/so/log/logstash/logstash.log` for clues around backoffs, etc. 

You could also use `so-logstash-pipeline-stats manager` (from the manager) to check the duration of each operation in the pipeline to see which operations are taking the longest.

**Reply by weslambert** [reply]
If the `tcp` output still ends up being too slow after confirming no issues with the SIEM, you may consider using the `udp` or `http` output and configuring `pool_max` to allow for a greater number of concurrent connections.

**Reply by bit1290** [reply]
Yes, I've checked any possible issues on the SIEM side but everything is working properly. The queues are fine.
After executing "so-logstash-pipeline-stats manager" I got this result:

[logstash_pipeline_stats.txt](https://github.com/Security-Onion-Solutions/securityonion/files/7719434/logstash_pipeline_stats.txt)

And I got this from "/opt/so/log/logstash/logstash.log":

[logstash.log](https://github.com/Security-Onion-Solutions/securityonion/files/7719445/logstash.log)

**Reply by bit1290** [reply]
I tried with udp but they are still delaying. 
Is there someway to optimize suricata?

**Reply by weslambert** [reply]
There are lots of ways to optimize Suricata, but if the same events are being fed into Security Onion on time, this particular issue likely has more to do with the Logstash pipeline or communication between/processing of the destination system, so I would troubleshoot and make adjustments there (unless this is being miscommunicated and the same events aren't actually there, but a subset are).

I would advise tuning Suricata regardless of the sending issue to greatly reduce the size/volume of your logs (assuming you aren't performing Suricata metadata collection), I would consider performing heavy tuning around the alerts it is generating, as ~130GB for each eve.json file is rather large.

**Reply by weslambert** [reply]
If it comes down to it being an issue with not being able to read the log files quickly enough, consider increasing the number of Filebeat workers.

**Reply by bit1290** [reply]
I just want alert type event to go to the eve.json files.
But suricata is not replicating the /opt/so/default/salt/suricata/defaults.yaml to /opt/so/conf/suricata/suricata.yaml

This is my defaults file:


[defaults (2).txt](https://github.com/Security-Onion-Solutions/securityonion/files/7721717/defaults.2.txt)

and this is my suricata.conf file:

[suricata (2).txt](https://github.com/Security-Onion-Solutions/securityonion/files/7721718/suricata.2.txt)


It justs keeps writing event types flow and DNS to the eve,json.

**Reply by weslambert** [reply]
Did you previously modify the config? Did you select to have Suricata or Zeek perform metadata collection during setup? What is the output of the following?

grep mdengine /opt/so/saltstack/local/pillar/global.sls

**Reply by bit1290** [reply]
Hello,

Q: Did you previously modify the config?
A: Yes I have

Q: Did you select to have Suricata or Zeek perform metadata collection during setup?
A: I selected Suricata

Q: What is the output of the following?
A: The output is the following:

[root@papidss003 ~]# grep mdengine /opt/so/saltstack/local/pillar/global.sls
  mdengine: 'SURICATA'
[root@papidss003 ~]#

Thanks for your reply!!
